sessiondata: fix ListSessions with SET LOCAL / SET ROLE

otan opened the pull request:
Previously, two incorrect phenomena was incorrect due to recent changes:
* SET ROLE would report the user the current session was impersonated
  as, rather than the user who was logged in. The latter seems more
  correct to report here.
* There is a race between a BEGIN/COMMIT/ROLLBACK and listing all
  sessions using ListSessions as they both try to access the stack.

This is both fixed by:
* Changing the returned user of SET ROLE to be the SessionUser.
* Introducing a `base` method to SessionDataStack, which returns the
  base of the stack. This is *always* set, so allow accordingly.

Release justification: fix to new functionality

Resolves: #69508

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/69553)
<!-- Reviewable:end -->

rafiss commented on pkg/sql/sessiondata/session_data.go:
```diff
@@ -229,6 +233,12 @@ func (s *Stack) Top() *SessionData {
 }
 
+// Base returns the bottom element of the stack.
+// This is a non-racy structure, as the bottom element is always constant.
```
very smart!

rafiss commented on pkg/sql/logictest/testdata/logic_test/set_role:
```diff
@@ -338,10 +340,28 @@ SELECT current_user, session_user()
 root  root
+
+statement ok
+SET LOCAL ROLE testuser
```
isn't this a no-op anyway since it's a SET LOCAL outside of a txn? or was it intended to be SET ROLE? (if so, maybe we might as well add a test for SET LOCAL ROLE inside a txn as well)

otan replied:
d'oh, yep!

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69553)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/logictest/testdata/logic_test/set_role, line 358 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69553#-MiJ2qzkvH8SAZiqOLn_-r1-358:-MiJofOy7k0km-LTqY4l:b-du8vbh) ([raw file](https://github.com/cockroachdb/cockroach/blob/6da24207e6b5e87dd14773bfce04e81f74be0a84/pkg/sql/logictest/testdata/logic_test/set_role#L358)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

d'oh, yep!
</blockquote></details>

ah oops i didn't see that SET LOCAL ROLE followed by `SELECT ... node_sessions` is already tested on line 336!

so i think the test case here (line 359) should be a normal SET ROLE 


<!-- Sent from Reviewable.io -->

otan committed:
sessiondata: fix ListSessions with SET LOCAL / SET ROLE

Previously, two incorrect phenomena was incorrect due to recent changes:
* SET ROLE would report the user the current session was impersonated
  as, rather than the user who was logged in. The latter seems more
  correct to report here.
* There is a race between a BEGIN/COMMIT/ROLLBACK and listing all
  sessions using ListSessions as they both try to access the stack.

This is both fixed by:
* Changing the returned user of SET ROLE to be the SessionUser.
* Introducing a `base` method to SessionDataStack, which returns the
  base of the stack. This is *always* set, so allow accordingly.

Release justification: fix to new functionality

Release note: None

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69553)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[pkg/sql/logictest/testdata/logic_test/set_role, line 358 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69553#-MiJ2qzkvH8SAZiqOLn_-r1-358:-MiK9O-S0IP04EdjwcvD:b-qkvsni) ([raw file](https://github.com/cockroachdb/cockroach/blob/6da24207e6b5e87dd14773bfce04e81f74be0a84/pkg/sql/logictest/testdata/logic_test/set_role#L358)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

ah oops i didn't see that SET LOCAL ROLE followed by `SELECT ... node_sessions` is already tested on line 336!

so i think the test case here (line 359) should be a normal SET ROLE 
</blockquote></details>

i managed to confuse myself, haha! fixed


<!-- Sent from Reviewable.io -->

otan commented:
bors r=rafiss

otan commented:
bors r=rafiss
